layers/ip6.go

```diff
@@ -138,21 +268,57 @@ func (i *IPv6) NextLayerType() gopacket.LayerType {
 func decodeIPv6(data []byte, p gopacket.PacketBuilder) error {
 	ip6 := &IPv6{}
 	err := ip6.DecodeFromBytes(data, p)
+	if err != nil {
+		return err
+	}
 	p.AddLayer(ip6)
 	p.SetNetworkLayer(ip6)
 	if ip6.HopByHop != nil {
-		// TODO(gconnell):  Since HopByHop is now an integral part of the IPv6
-		// layer, should it actually be added as its own layer?  I'm leaning towards
-		// no.
 		p.AddLayer(ip6.HopByHop)
+		return p.NextDecoder(ip6.HopByHop.NextHeader)
 	}
+	return p.NextDecoder(ip6.NextHeader)
+}
+
+func (i *IPv6HopByHop) serialize(opts gopacket.SerializeOptions) ([]byte, error) {
+	b := gopacket.NewSerializeBuffer()
+
+	o := make([]ipv6HeaderTLVOption, 0, len(i.Options))
+	for _, v := range i.Options {
+		o = append(o, ipv6HeaderTLVOption(v))
+	}
+	err := serializeIPv6HeaderTLVOptions(b, o, opts.FixLengths)
+	if err != nil {
+		return nil, err
+	}
+	hbhopts := b.Bytes()
+	length := len(hbhopts) + 2
+	if length%8 != 0 {
+		return nil, fmt.Errorf("IPv6HopByHop actual length must be multiple of 8")
+	}
+	bytes, err := b.PrependBytes(2)
+	if err != nil {
+		return nil, err
+	}
+	bytes[0] = uint8(i.NextHeader)
+	if opts.FixLengths {
+		i.HeaderLength = uint8((length / 8) - 1)
+	}
+	bytes[1] = uint8(i.HeaderLength)
+	return b.Bytes(), nil
+}
+
+func (i *IPv6HopByHop) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.SerializeOptions) error {
+	buf, err := i.serialize(opts)
```

Comment:
One would have to precompute the serialization front to back first because of the padding... then prepend the final amount of bytes. Then copy front to back. Is it worth it? I have no problem working on it.

I think we should be strict about adhering to the supplied TLV ordering in IPv6HopByHop.Options. Do you agree?